### PR TITLE
Add channel pool for remote execution to overcome gRPC connections limitation.

### LINKED ISSUE #11801: Remote concurrency limited by gRPC connections
### Description of the problem / feature request:

For remote RPCs, bazel only opens one connection per server (resolved IP). This constrains the maximum parallelism of a build - gRPC limits the number of in-flight requests per connection to MAX_CONCURRENT_STREAMS, which is a per-server HTTP2 setting, often set to 100 or 128.

In short, this means that when talking to a server resolved to a single IP, bazel can only have ~100 remote executions in flight, regardless of the setting of --jobs. If the server resolves to more IPs, this may be linearly increased, but rarely above 500. Most remote builds can exceed 100x parallelism, and many (esp. builds with long-duration actions) could exceed 500.

Note that due to bazel _displaying_ full parallelism even if actions are actually queued waiting for gRPC connections, it's hard to know just how many users this affects - more than realize it, I'm quite sure. But I've had at least 3 independent projects root cause this reason for limited parallelism at this point, so it seems to be fairly frequent in reality (at least amongst RBE users).

The necessary fix in bazel is to replace the netty 'round_robin' load balancer with one that can open more than 1 connection (subchannel) per resolved address. Maybe by dynamically adding more when bottlenecked, but feeding in a target number of connections as e.g. `max(2, jobs/50)` would probably do just fine.

### Feature requests: what underlying problem are you trying to solve with this feature?

Enable higher build parallelism for remote builds

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Run a remote build with --jobs=1000 and observe bazel showing '1000 remote'. Then check server perspective and realize bazel never had more than X00 RPCs in flight. 

### What operating system are you running Bazel on?

Linux (but applies to all)

### What's the output of `bazel info release`?

3.4.1

###  Have you found anything relevant by searching the web?

 - gRPC are not currently planning on offering this feature natively: https://github.com/grpc/grpc/issues/21386
 - Usually resolved in clients by writing a custom load balancer, as is likely the necessary fix for bazel. Example: https://github.com/GoogleCloudPlatform/grpc-gcp-go/tree/master/grpcgcp

cc @buchgr @bergsieker @gkousik @ulfjack

### PR DESCRIPTION
**Pull request by coeuvre**
This PR add a `ReferenceCountedChannelPool` which will create `poolSize` number of channels and round-robin across them for gRPC requests.

The `poolSize` is calculated as `jobs / 100`.

Fixes #11801.

### CODE REVIEW

**Comment by coeuvre** [comment]
cc @ulfjack

File: src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java [hunk 1]
```diff
@@ -255,6 +256,13 @@ public void beforeCommand(CommandEnvironment env) throws AbruptExitException {
     ReferenceCountedChannel execChannel = null;
     ReferenceCountedChannel cacheChannel = null;
     ReferenceCountedChannel downloaderChannel = null;
+
+    int poolSize = 1;
+    BuildRequestOptions buildRequestOptions = env.getOptions().getOptions(BuildRequestOptions.class);
+    if (buildRequestOptions != null) {
+      // TODO(chiwang): Should we use a more intelligent way to calculate poolSize?
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java:263]: yeah, this is my main question: How did you come up with this number?

Comment (gkousik) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java:263]: I believe its based on the recommendation here - https://github.com/bazelbuild/bazel/issues/11801#issuecomment-672973245

Comment (meisterT) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java:263]: That makes sense indeed. Chi, can you please document this in the code?

File: src/main/java/com/google/devtools/build/lib/remote/ReferenceCountedChannelPool.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
+package com.google.devtools.build.lib.remote;
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/ReferenceCountedChannelPool.java:1]: Please add the default copyright header here

File: src/main/java/com/google/devtools/build/lib/remote/ReferenceCountedChannelPool.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  @Override
+  public String authority() {
+    // Assume all channels have the same authority.
+    return channels.get(0).authority();
+  }
+
+  /**
+   * Performs a simple round robin on the list of {@link ManagedChannel}s in the {@code channels}
+   * list.
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/ReferenceCountedChannelPool.java:84]: Reference https://github.com/grpc/grpc/issues/21386#issuecomment-564742173 in the comment?

**Comment by meisterT** [comment]
@coeuvre ISTR that Bazel has a RBE pool with 100 executors (I might be mistaken), but I would like you to test this before submission. Perhaps @EricBurnett can help you out with a larger pool.

**Comment by gkousik** [comment]
I just tested it using repro I had previously here - https://github.com/gkousik/bazel-parallelism-test. It simply runs 490 actions in parallel where each action sleeps for 10 seconds. 

With this change (whole build runs in 12s on a 500 worker pool):
```
➜  bazel-parallelism-test git:(master) ✗ ./runbuild.sh
INFO: Invocation ID: eb16917c-3c40-4407-8e58-f2c3973e9e5c
ERROR: /usr/local/google/home/kousikk/.cache/bazel/_bazel_kousikk/07c64333bc485db1757df85f8da29b91/external/bazel_gazelle_go_repository_config/WORKSPACE:3:1: name 'go_repository' is not defined (did you mean 'local_repository'?)
ERROR: /usr/local/google/home/kousikk/.cache/bazel/_bazel_kousikk/07c64333bc485db1757df85f8da29b91/external/bazel_gazelle_go_repository_config/WORKSPACE:8:1: name 'go_repository' is not defined (did you mean 'local_repository'?)
DEBUG: /usr/local/google/home/kousikk/.cache/bazel/_bazel_kousikk/07c64333bc485db1757df85f8da29b91/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:59:14:
Current running Bazel is not a release version and one was not defined explicitly in rbe_autoconfig target. Falling back to '3.2.0'
INFO: Build option --action_env has changed, discarding analysis cache.
DEBUG: /usr/local/google/home/kousikk/.cache/bazel/_bazel_kousikk/07c64333bc485db1757df85f8da29b91/external/bazel_toolchains/rules/rbe_repo.bzl:491:10: Bazel 3.2.0 is used in rbe_default.
INFO: Analyzed 489 targets (6 packages loaded, 500 targets configured).
INFO: Found 489 targets...
INFO: Elapsed time: 12.484s, Critical Path: 11.87s
INFO: 489 processes: 489 remote.
INFO: Build completed successfully, 490 total actions
```

Without this change (whole build takes 35s on a 500 worker pool):
```
➜  bazel-parallelism-test git:(master) ✗ ./runbuild.sh
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 5e7bc63c-3e9f-42c5-95c4-300a444c4ddb
DEBUG: /usr/local/google/home/kousikk/.cache/bazel/_bazel_kousikk/07c64333bc485db1757df85f8da29b91/external/bazel_toolchains/rules/rbe_repo.bzl:491:10: Bazel 3.2.0 is used in rbe_default.
INFO: Analyzed 489 targets (7 packages loaded, 500 targets configured).
INFO: Found 489 targets...
INFO: Elapsed time: 53.713s, Critical Path: 35.42s
INFO: 489 processes: 489 remote.
INFO: Build completed successfully, 490 total actions
```

**Comment by coeuvre** [comment]
@meisterT  Yes, I have tested before create this PR and got similar result. Thanks for your comments!

File: src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java (hunk 1 above)
Comment (EricBurnett) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java:263]: Note that the recommendation there is to either take into account the number of IPs (the number of connections, aka subchannels, per channel), or to assume 2 IPs per host which would be jobs/100 "Channels" in this pool.

Looking at this code, just assuming 2 makes more sense to me than having to also reimplement the subchannel->channel interface, which I think would be required to count the IPs. So I vote for simply changing this constant to 100 and documenting clearly why it is what it is.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java [hunk 4]
```diff
@@ -255,6 +256,22 @@ public void beforeCommand(CommandEnvironment env) throws AbruptExitException {
...
+      // The following calculation is based on the suggestion from comment
+      // https://github.com/bazelbuild/bazel/issues/11801#issuecomment-672973245
+      //
+      // The number of concurrent requests for one connection to a gRPC server is limited by
+      // MAX_CONCURRENT_STREAMS which is normally being 100+. We assume 50 concurrent requests for
+      // each connection should be fairly well. The number of connections opened by one channel is
+      // based on the resolved IPs of that server. We assume servers normally have 2 IPs. So the
+      // number of required channel is calculated as: ceil(jobs / 100).
+      //
+      // TODO(chiwang): Should we use a more intelligent way to calculate poolSize?
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java:272]: I think you can remove this TODO.

**Pull request closed without merging**